manticore/core/memory.py

```diff
@@ -1045,6 +1085,74 @@ def _try_get_solutions(self, address, size, access, max_solutions=0x1000, force=
         return solutions
 
 
+class LazySMemory(SMemory):
+    '''
+    A fully symbolic memory.
+
+    Currently does not support cross-page reads/writes.
+    '''
+    def __init__(self, constraints, *args, **kwargs):
+        super(LazySMemory, self).__init__(constraints, *args, **kwargs)
+
+    def mmap(self, addr, size, perms, name=None, **kwargs):
+        assert addr is None or isinstance(addr, (int, long)), 'Address must be concrete'
+        assert addr < self.memory_size, 'Address too big'
+        assert size > 0
+
+        if addr is not None:
+            addr = self._floor(addr)
+
+        size = self._ceil(size)
+
+        addr = self._search(size, addr)
+
+        # It should not be allocated
+        for i in xrange(self._page(addr), self._page(addr + size)):
+            assert i not in self._page2map, 'Map already used'
```

Comment:
we should use an exception, not an assert here probably, in case we run w/ asserts disabled